File: autozig-math/src/dmat4.rs
use autozig::include_zig;
use crate::{DVec3, DVec4};

#[repr(C)]
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct DMat4 {
    pub cols: [[f64; 4]; 4],
}

include_zig!("zig/dmat4.zig", {
    fn dmat4_identity() -> DMat4;
    fn dmat4_from_translation(translation: DVec3) -> DMat4;
    fn dmat4_from_scale(scale: DVec3) -> DMat4;
    fn dmat4_mul(self_: DMat4, other: DMat4) -> DMat4;
    fn dmat4_mul_dvec4(self_: DMat4, vec: DVec4) -> DVec4;
    fn dmat4_transpose(self_: DMat4) -> DMat4;
    fn dmat4_transform_point(self_: DMat4, point: DVec3) -> DVec3;
});

impl DMat4 {
    pub const IDENTITY: Self = Self {
        cols: [
            [1.0, 0.0, 0.0, 0.0],
            [0.0, 1.0, 0.0, 0.0],
            [0.0, 0.0, 1.0, 0.0],
            [0.0, 0.0, 0.0, 1.0],
        ],
    };

    pub fn identity() -> Self {
        dmat4_identity()
    }

    pub fn from_translation(translation: DVec3) -> Self {
        dmat4_from_translation(translation)
    }

    pub fn from_scale(scale: DVec3) -> Self {
        dmat4_from_scale(scale)
    }

    pub fn mul_dvec4(self, vec: DVec4) -> DVec4 {
        dmat4_mul_dvec4(self, vec)
    }

    pub fn transpose(self) -> Self {
        dmat4_transpose(self)
    }

    pub fn transform_point(self, point: DVec3) -> DVec3 {
        dmat4_transform_point(self, point)
    }
}

impl std::ops::Mul for DMat4 {
    type Output = Self;
    fn mul(self, rhs: Self) -> Self::Output {
        dmat4_mul(self, rhs)
    }
}

impl std::ops::Mul<DVec4> for DMat4 {
    type Output = DVec4;
    fn mul(self, rhs: DVec4) -> Self::Output {
        dmat4_mul_dvec4(self, rhs)
    }
}

File: autozig-math/zig/dmat4.zig
const std = @import("std");
const DVec3 = @import("dvec3.zig").DVec3;
const DVec4 = @import("dvec4.zig").DVec4;
const DQuat = @import("dquat.zig").DQuat;

pub const DMat4 = extern struct {
    cols: [4][4]f64,

    pub const IDENTITY = DMat4{ .cols = .{
        .{ 1.0, 0.0, 0.0, 0.0 },
        .{ 0.0, 1.0, 0.0, 0.0 },
        .{ 0.0, 0.0, 1.0, 0.0 },
        .{ 0.0, 0.0, 0.0, 1.0 },
    } };

    pub const ZERO = DMat4{ .cols = .{
        .{ 0.0, 0.0, 0.0, 0.0 },
        .{ 0.0, 0.0, 0.0, 0.0 },
        .{ 0.0, 0.0, 0.0, 0.0 },
        .{ 0.0, 0.0, 0.0, 0.0 },
    } };

    pub fn identity() DMat4 {
        return IDENTITY;
    }

    pub fn from_translation(translation: DVec3) DMat4 {
        return .{ .cols = .{
            .{ 1.0, 0.0, 0.0, 0.0 },
            .{ 0.0, 1.0, 0.0, 0.0 },
            .{ 0.0, 0.0, 1.0, 0.0 },
            .{ translation.x, translation.y, translation.z, 1.0 },
        } };
    }

    pub fn from_scale(scale: DVec3) DMat4 {
        return .{ .cols = .{
            .{ scale.x, 0.0, 0.0, 0.0 },
            .{ 0.0, scale.y, 0.0, 0.0 },
            .{ 0.0, 0.0, scale.z, 0.0 },
            .{ 0.0, 0.0, 0.0, 1.0 },
        } };
    }

    pub fn from_dquat(q: DQuat) DMat4 {
        const x2 = q.x + q.x;
        const y2 = q.y + q.y;
        const z2 = q.z + q.z;
        const xx = q.x * x2;
        const xy = q.x * y2;
        const xz = q.x * z2;
        const yy = q.y * y2;
        const yz = q.y * z2;
        const zz = q.z * z2;
        const wx = q.w * x2;
        const wy = q.w * y2;
        const wz = q.w * z2;

        return .{ .cols = .{
            .{ 1.0 - (yy + zz), xy + wz, xz - wy, 0.0 },
            .{ xy - wz, 1.0 - (xx + zz), yz + wx, 0.0 },
            .{ xz + wy, yz - wx, 1.0 - (xx + yy), 0.0 },
            .{ 0.0, 0.0, 0.0, 1.0 },
        } };
    }

    pub fn mul(self: DMat4, other: DMat4) DMat4 {
        var result: DMat4 = undefined;
        inline for (0..4) |col| {
            inline for (0..4) |row| {
                var sum: f64 = 0.0;
                inline for (0..4) |k| {
                    sum += self.cols[k][row] * other.cols[col][k];
                }
                result.cols[col][row] = sum;
            }
        }
        return result;
    }

    pub fn mul_dvec4(self: DMat4, vec: DVec4) DVec4 {
        return .{
            .x = self.cols[0][0] * vec.x + self.cols[1][0] * vec.y + self.cols[2][0] * vec.z + self.cols[3][0] * vec.w,
            .y = self.cols[0][1] * vec.x + self.cols[1][1] * vec.y + self.cols[2][1] * vec.z + self.cols[3][1] * vec.w,
            .z = self.cols[0][2] * vec.x + self.cols[1][2] * vec.y + self.cols[2][2] * vec.z + self.cols[3][2] * vec.w,
            .w = self.cols[0][3] * vec.x + self.cols[1][3] * vec.y + self.cols[2][3] * vec.z + self.cols[3][3] * vec.w,
        };
    }

    pub fn transpose(self: DMat4) DMat4 {
        return .{ .cols = .{
            .{ self.cols[0][0], self.cols[1][0], self.cols[2][0], self.cols[3][0] },
            .{ self.cols[0][1], self.cols[1][1], self.cols[2][1], self.cols[3][1] },
            .{ self.cols[0][2], self.cols[1][2], self.cols[2][2], self.cols[3][2] },
            .{ self.cols[0][3], self.cols[1][3], self.cols[2][3], self.cols[3][3] },
        } };
    }

    pub fn transform_point(self: DMat4, point: DVec3) DVec3 {
        return .{
            .x = self.cols[0][0] * point.x + self.cols[1][0] * point.y + self.cols[2][0] * point.z + self.cols[3][0],
            .y = self.cols[0][1] * point.x + self.cols[1][1] * point.y + self.cols[2][1] * point.z + self.cols[3][1],
            .z = self.cols[0][2] * point.x + self.cols[1][2] * point.y + self.cols[2][2] * point.z + self.cols[3][2],
        };
    }
};

export fn dmat4_identity() DMat4 {
    return DMat4.identity();
}

export fn dmat4_from_translation(translation: DVec3) DMat4 {
    return DMat4.from_translation(translation);
}

export fn dmat4_from_scale(scale: DVec3) DMat4 {
    return DMat4.from_scale(scale);
}

export fn dmat4_mul(self: DMat4, other: DMat4) DMat4 {
    return self.mul(other);
}

export fn dmat4_mul_dvec4(self: DMat4, vec: DVec4) DVec4 {
    return self.mul_dvec4(vec);
}

export fn dmat4_transpose(self: DMat4) DMat4 {
    return self.transpose();
}

export fn dmat4_transform_point(self: DMat4, point: DVec3) DVec3 {
    return self.transform_point(point);
}
